engine/src/main/java/org/hibernate/search/engine/search/query/spi/QueryParametersValueProvider.java

```diff
@@ -0,0 +1,108 @@
+		return new SimpleValueProvider<>( value );
+	}
+
+	static <F> QueryParametersValueProvider<F> parameter(String parameterName, Class<F> parameterType) {
+		return new ParameterValueProvider<>( parameterName, parameterType, Function.identity() );
+	}
+
+	static <F, R> QueryParametersValueProvider<F> parameter(String parameterName, Class<? extends R> parameterType,
+			Function<? super R, ? extends F> mapping) {
+		return new ParameterValueProvider<>( parameterName, parameterType, mapping );
+	}
+
+	static <F, R> QueryParametersValueProvider<F> parameterCollectionOrSingle(String parameterName,
+			Function<Collection<?>, ? extends F> mapping) {
+		return new ParameterCollectionOrSingleValueProvider<>( parameterName, mapping );
+	}
+
+	static QueryParametersValueProvider<Double> distanceInMeters(String radius, String unit) {
+		return new DistanceInMetersProvider( radius, unit );
+	}
+
+	class SimpleValueProvider<F> implements QueryParametersValueProvider<F> {
+		private final F value;
+
+		public SimpleValueProvider(F value) {
+			this.value = value;
+		}
+
+		@Override
+		public F provide(QueryParametersContext context) {
+			return value;
+		}
+	}
+
+	class ParameterValueProvider<F, R> implements QueryParametersValueProvider<F> {
+		private final String parameterName;
+		private final Class<? extends R> parameterType;
+		private final Function<? super R, ? extends F> mapping;
+
+		public ParameterValueProvider(String parameterName, Class<? extends R> parameterType,
+				Function<? super R, ? extends F> mapping) {
+			this.parameterName = parameterName;
+			this.parameterType = parameterType;
+			this.mapping = mapping;
+		}
+
+		@Override
+		public F provide(QueryParametersContext context) {
+			return mapping.apply( context.parameter( parameterName, parameterType ) );
+		}
+	}
+
+	class DistanceInMetersProvider implements QueryParametersValueProvider<Double> {
+
+		private final String radius;
+		private final String unit;
+
+		public DistanceInMetersProvider(String radius, String unit) {
+			this.radius = radius;
+			this.unit = unit;
+		}
+
+		@Override
+		public Double provide(QueryParametersContext context) {
+			DistanceUnit unitValue = context.parameter( unit, DistanceUnit.class );
+			Double radiusValue = context.parameter( radius, Double.class );
+			return unitValue.toMeters( radiusValue );
+		}
+	}
+
+	class ParameterCollectionOrSingleValueProvider<F> implements QueryParametersValueProvider<F> {
+		private final String parameterName;
+		private final Function<Collection<?>, ? extends F> mapping;
+
+		public ParameterCollectionOrSingleValueProvider(String parameterName, Function<Collection<?>, ? extends F> mapping) {
+			this.parameterName = parameterName;
+			this.mapping = mapping;
+		}
+
+		@Override
+		public F provide(QueryParametersContext context) {
+			Object parameter = context.parameter( parameterName );
+			return mapping.apply( parameter instanceof Collection ? (Collection<?>) parameter : List.of( parameter ) );
+		}
```

Comment:
not 100% on this one, it is to address the case of various terms predicates,  those predicates expect 1-n terms
and with our current DSL user can use a vararg and pass a single term or an array/list of them. But with a parameter, there's no way to tell what an expectation is, so the idea is that we always want to get a collection, and if the user gives a single element, then we wrap it into a list.